### ISSUE TITLE
systemd.network(5) Type=ether does not match due to DEVTYPE not being set by udev

### ISSUE THREAD

**Issue by Neurognostic** [systemd/systemd#14952]
**systemd version the issue has been seen with**
> systemd 244 (244.3-1-arch)

**Used distribution**
 > Arch Linux

**Expected behaviour you didn't see**
 > Need to create a default network configuration for all Ethernet devices (similar to #3998).

**Unexpected behaviour you saw**
 > `systemd.network(5)` files using `Type=ether` do not match Ethernet devices.

**Steps to reproduce the problem**

```shell
cat<<EOF>/etc/systemd/network/00-ether-test.network
[Match]
Type=ether
[Network]
DHCP=ipv4
EOF

networkctl reload
```

# Summary

When specifying `Type=ether` in `systemd.network(5)`, Ethernet devices are not matched. I believe this is because `udev(7)` does not define a `DEVTYPE`  property for Ethernet devices. However, when running `networkctl status`, `networkctl(1)` correctly identifies the device as `Type: ether`.

## How the issue was verified

Quote from `systemd.network(5)`:
>**Type=**
>           A whitespace-separated list of shell-style globs matching
>           the device type, as exposed by the udev property
>           "DEVTYPE". If the list is prefixed with a "!", the test
>           is inverted.

`udevadm info` shows that no `DEVTYPE` property is set.

```console
# udevadm info /sys/class/net/enp0s31f6
P: /devices/pci0000:00/0000:00:1f.6/net/enp0s31f6
L: 0
E: DEVPATH=/devices/pci0000:00/0000:00:1f.6/net/enp0s31f6
E: INTERFACE=enp0s31f6
E: IFINDEX=2
E: SUBSYSTEM=net
E: USEC_INITIALIZED=22963369
E: ID_NET_NAMING_SCHEME=v243
E: ID_NET_NAME_MAC=enx482ae3fa5e16
E: ID_OUI_FROM_DATABASE=Wistron InfoComm(Kunshan)Co.,Ltd.
E: ID_NET_NAME_PATH=enp0s31f6
E: ID_BUS=pci
E: ID_VENDOR_ID=0x8086
E: ID_MODEL_ID=0x15bc
E: ID_PCI_CLASS_FROM_DATABASE=Network controller
E: ID_PCI_SUBCLASS_FROM_DATABASE=Ethernet controller
E: ID_VENDOR_FROM_DATABASE=Intel Corporation
E: ID_MODEL_FROM_DATABASE=Ethernet Connection (7) I219-V
E: ID_PATH=pci-0000:00:1f.6
E: ID_PATH_TAG=pci-0000_00_1f_6
E: ID_NET_DRIVER=e1000e
E: ID_NET_LINK_FILE=/usr/lib/systemd/network/99-default.link
E: SYSTEMD_ALIAS=/sys/subsystem/net/devices/enp0s31f6
E: TAGS=:systemd:
```

`networkctl status` shows `Type: ether`, however, no `Network File:` is matching  with the above `systemd.network(5)` file.

```console
# networkctl status enp0s31f6
● 2: enp0s31f6
             Link File: /usr/lib/systemd/network/99-default.link
          Network File: n/a
                  Type: ether
                 State: routable (unmanaged)
                  Path: pci-0000:00:1f.6
                Driver: e1000e
                Vendor: Intel Corporation
                 Model: Ethernet Connection (7) I219-V
            HW Address: 48:2a:e3:fa:5e:16 (Wistron InfoComm(Kunshan)Co.,Ltd.)
                   MTU: 1500 (min: 68, max: 9000)
  Queue Length (Tx/Rx): 1/1
      Auto negotiation: yes
                 Speed: 1Gbps
                Duplex: full
                  Port: tp
               Address: 10.0.1.130
                        fe80::4a2a:e3ff:fe56:2a55
               Gateway: 10.0.1.1 (NETGEAR)
```

### Logs

Debug udev logs, if needed.

```shell
SYSTEMD_LOG_LEVEL=debug udevadm test /sys/class/net/enp0s31f6
```
Output attached: [udevadm.test.debug.txt](https://github.com/systemd/systemd/files/4259683/udevadm.test.debug.txt)

## Possible Workarounds

These probably do not work in all cases, but they do in my current use cases.

```shell
cat<<EOF>/etc/systemd/network/00-ether-test.network
[Match]
Name=e*
[Network]
DHCP=ipv4
EOF
```

_Or_

```shell
cat<<EOF>/etc/systemd/network/00-ether-test.network
[Match]
Property="ID_PCI_SUBCLASS_FROM_DATABASE=Ethernet controller"
[Network]
DHCP=ipv4
EOF
```

## Other thoughts

Is it possible to use whatever `networkctl status` is using to determine `Type: ether` in addition to the `udev(7)` property `DEVTYPE`?

https://github.com/systemd/systemd/issues/3998#issuecomment-251674635 mentions possible solutions.

**Comment by Neurognostic** [comment]
https://lists.freedesktop.org/archives/systemd-devel/2014-January/015999.html

>Hi Greg,
>
>>>>> I just pushed a change[0] which allows the match syntax
>>>>> "Type=ethernet" to match on network devices without a DEVTYPE.
>>>>> 
>>>>> We had a discussion on IRC whether we should call it Type=wired or
>>>>> Type=ethernet. I think the former may be more intuitive, but the
>>>>> latter seems to be more in line with what is done elsewhere (connman
>>>>> calls it ethernet, and udev prefixes the devices names with 'en').
>>>>> 
>>>>> Any thoughts?
>>>> 
>>>> Any reason why the kernel can't be setting this value in the first place
>>>> so you don't have to rely on it being "null"?
>>>> 
>>>> Seems like a kernel issue.
>>> 
>>> I asked Marcel the same thing earlier. He said:
>>> 
>>> [Friday 03 January 2014] [19:58:22] <holtmann>  Because you have to
>>> touch every single driver to get this done properly.
>>> [Friday 03 January 2014] [19:58:34] <holtmann>  So DEVTYPE= means it
>>> is wired Ethernet.
>>> [Friday 03 January 2014] [19:58:42] <holtmann>  If that is not true,
>>> then that is a bug.
>>> 
>>> From my point of view, I'd prefer the kernel just doing the right
>>> thing (and I'd be happy to write the patches), but my impression was
>>> that that's not going to happen...(?)
>> 
>> It shouldn't need to be set for every driver, wireless doesn't work this
>> way, so there is lots of precident for how to do this in only a few
>> lines of kernel code :)
>
>when I did the initial DEVTYPE support, it was easy for “wlan” since the subsystem is taking care of creating the netdev. So you have one places for mac80211 drivers and some extra work for the cfg80211 only drivers. It was all minimal work. So was “bluetooth” since it is one central place and same applied to “wimax”. The “wwan” was lucky part since the drivers already did the FLAG_WWAN to pick the wwan%d ifname.
>
>For Ethernet cards it was just the wild west. So I left them out of the picture. It is also more important to detect the devices that need extra management handling. Like WiFi, Bluetooth, 3G/LTE etc. The indication of DEVTYPE here is mainly that nobody goes and tries to run DHCP until another management entity tell you the device is ready. For Ethernet or DEVTYPE=“” you do not need to know who manages it since there is nobody.
>
>Regards
>
>Marcel

According to this, when `DEVTYPE` is not set it is always a wired/Ethernet device.

**Comment by yuwata** [comment]
Fix is waiting in #14957.
